src/ext/service.rs

```diff
@@ -99,65 +98,69 @@ impl ServiceExt for dc::Service {
-                           (&label[prefix.len()..]).to_owned()));
+                let lib_name = label[prefix.len()..].to_string();
+                let source: Result<&Source> = sources
+                    .find_by_lib_key(&lib_name)
+                    .ok_or_else(|| ErrorKind::UnknownLibKey(lib_name).into());
+
+                libs.push(SourceMount {
+                    container_path: mount_as.value()?.to_owned(),
+                    source: source?,
+                    source_subdirectory: None,
+                })
             }
         }
 
-        Ok(Sources {
-            sources: sources,
-            context: context,
-            libs: libs.into_iter(),
-        })
+        Ok(Sources { context, libs: libs.into_iter() })
     }
 }
 
 /// Iterator over all the `Source` trees which can be mounted into this
 /// `Service`.
 pub struct Sources<'a> {
-    /// All `Source` trees available for this repository.
-    sources: &'a sources::Sources,
     /// Do we need to iterate over our `context` field?
-    context: Option<(String, Option<String>, dc::Context)>,
+    context: Option<SourceMount<'a>>,
     /// Libraries
-    libs: vec::IntoIter<(String, String)>,
+    libs: vec::IntoIter<SourceMount<'a>>,
+}
+
+#[derive(Clone)]
+pub struct SourceMount<'a> {
+    pub container_path: String,
+    pub source: &'a Source,
+    pub source_subdirectory: Option<String>,
 }
 
 impl<'a> Iterator for Sources<'a> {
-    type Item = Result<(String, Option<String>, &'a Source)>;
+    type Item = SourceMount<'a>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        // Check for a `dc::Context` using `take`, which moves data out of
+        // Check for a `SourceMount` using `take`, which moves data out of
         // an `Option` value and leaves `None` in its place,
         // simulataneously updating our internal state and keeping the
         // borrow checker happy.
-        if let Some((container_path, repo_subdir, context)) = self.context.take() {
-            if let Some(source) = self.sources.find_by_alias(&context.human_alias().unwrap()) {
-                // We have a `context` and a `source`, so return them.
-                Some(Ok((container_path, repo_subdir, source)))
-            } else {
-                // We have a `context` but it doesn't correspond to a known
-                // `Source`, so move on the next step of the iteration.
-                self.next()
-            }
-        } else {
-            // Iterate over any "libs"-style mounts.
-            self.libs.next().map(|(container_path, name)| {
-                match self.sources.find_by_lib_key(&name) {
-                    None => Err(ErrorKind::UnknownLibKey(name).into()),
-                    Some(source) => Ok((container_path, None, source)),
-                }
-            })
-        }
+        self.context.take().or_else(|| {
+             // If there is no SourceMount for the service itself, iterate over libs
+            self.libs.next()
+        })
```

Comment:
Iterator goes from 18 lines down to 4! 😄

Reply:
Yay!